File: src/qkb/search/retrieval.py
"""Document retrieval by UUID or prefix, three formats (DESIGN.md §8.8)."""

from __future__ import annotations

import sqlite3
from pathlib import Path

from qkb.search.results import context_description, hydrate

_LIKE_ESCAPE = "\\"


class DocumentFileMissing(FileNotFoundError):
    """Raised when a document's on-disk file is gone since the last ingest.

    Subclasses FileNotFoundError so callers that only expect the builtin
    still catch it, while giving CLI/MCP a specific type to match on for a
    friendly message.
    """


def _escape_like_prefix(raw: str) -> str:
    """Escape LIKE metacharacters so `raw` matches only literally, then
    append the trailing wildcard for prefix matching."""
    escaped = (
        raw.replace(_LIKE_ESCAPE, _LIKE_ESCAPE * 2)
        .replace("%", _LIKE_ESCAPE + "%")
        .replace("_", _LIKE_ESCAPE + "_")
    )
    return escaped + "%"


def get_document(
    conn: sqlite3.Connection,
    id_or_prefix: str,
    vault_path: Path | None = None,
    include_raw: bool = False,
    include_siblings: bool = True,
) -> dict:
    rows = conn.execute(
        f"SELECT id FROM documents WHERE id LIKE ? ESCAPE '{_LIKE_ESCAPE}'",
        (_escape_like_prefix(id_or_prefix),),
    ).fetchall()
    if not rows:
        raise KeyError(f"no document with id (prefix) {id_or_prefix!r}")
    if len(rows) > 1:
        raise ValueError(f"ambiguous prefix {id_or_prefix!r} matches {len(rows)} documents")
    doc = hydrate(conn, [(rows[0]["id"], 0.0, None)])[0]
    del doc["score"], doc["matched_text"]
    if not include_siblings:
        doc["siblings"] = []
    if include_raw:
        if vault_path is None:
            raise ValueError("include_raw requires vault_path")
        file_path = vault_path / doc["file_path"]
        try:
            text = file_path.read_text(encoding="utf-8")
        except FileNotFoundError as e:
            raise DocumentFileMissing(
                f"file moved or deleted since last ingest: {doc['file_path']!r} "
                f"(document {doc['document_id']!r}) — re-run `qkb ingest`"
            ) from e
        except OSError as e:
            # Below-the-cut: FileNotFoundError is the common case (moved/deleted
            # since last ingest) and gets the friendly message above. Anything
            # else OSError-shaped (PermissionError, IsADirectoryError, ...) still
            # means "can't read this document's file" and should surface as the
            # same typed, catchable error rather than a raw traceback.
            # (UnicodeDecodeError subclasses ValueError, not OSError, and is
            # already handled by callers expecting ValueError.)
            raise DocumentFileMissing(
                f"cannot read file {doc['file_path']!r} (document {doc['document_id']!r}): {e}"
            ) from e
        desc = context_description(conn, doc["context"])
        if desc:
            text = f"<!-- Context: {desc} -->\n\n{text}"
        doc["raw_text"] = text
    return doc

File: src/search/retrieval.ts
/** Document retrieval by UUID or prefix, three formats (DESIGN.md §8.8).
 * Ported from `legacy/python/src/qkb/search/retrieval.py`. */

import { readFileSync } from "node:fs";
import { join } from "node:path";
import type Database from "better-sqlite3";
import type { HydratedResult } from "./hydrate.js";
import { contextDescription, hydrate } from "./hydrate.js";

const LIKE_ESCAPE = "\\";

/** Raised when the id/prefix argument matches no document. Python raises the
 * builtin `KeyError` here; TS has no exact equivalent, so this is a distinct
 * class callers can `instanceof`-check. */
export class DocumentNotFoundError extends Error {}

/** Raised when a prefix matches more than one document. Python raises the
 * builtin `ValueError` here; see `DocumentNotFoundError` for why this is a
 * distinct TS class instead. */
export class AmbiguousDocumentPrefixError extends Error {}

/**
 * Raised when a document's on-disk file is gone or unreadable since the last
 * ingest (moved, deleted, replaced by a directory, permission denied, ...).
 *
 * Python's equivalent subclasses the builtin `FileNotFoundError` so callers
 * that only know about the builtin still catch it. Node has no comparably
 * distinguished builtin hierarchy for filesystem errors (readFileSync throws
 * a plain `Error` with an `.code` like `ENOENT`/`EISDIR` regardless of
 * cause), so this is a plain `Error` subclass — callers match on
 * `instanceof DocumentFileMissing` instead.
 */
export class DocumentFileMissing extends Error {}

/** Escape LIKE metacharacters so `raw` matches only literally, then append
 * the trailing wildcard for prefix matching. Ported from
 * `retrieval.py`'s `_escape_like_prefix`. */
function escapeLikePrefix(raw: string): string {
  const escaped = raw
    .replaceAll(LIKE_ESCAPE, LIKE_ESCAPE + LIKE_ESCAPE)
    .replaceAll("%", `${LIKE_ESCAPE}%`)
    .replaceAll("_", `${LIKE_ESCAPE}_`);
  return `${escaped}%`;
}

/** `hydrate`'s result shape minus the search-only `score`/`matched_text`
 * fields (a single document lookup isn't a ranked search hit), plus the raw
 * markdown text when `includeRaw` is set. */
export type DocumentDetail = Omit<HydratedResult, "score" | "matched_text"> & {
  raw_text?: string;
};

/**
 * Look up a document by exact id or unambiguous id prefix.
 *
 * Throws `DocumentNotFoundError` if no document matches, or
 * `AmbiguousDocumentPrefixError` if more than one does. When `includeRaw` is
 * set, also reads the note's file from `vaultPath` (read-only — never
 * writes) as raw UTF-8 text, prefixed with an HTML comment carrying the
 * context description if one exists; throws `DocumentFileMissing` if that
 * read fails for any reason. Ported from `retrieval.py`'s `get_document`.
 */
export function getDocument(
  conn: Database.Database,
  idOrPrefix: string,
  vaultPath?: string,
  includeRaw = false,
  includeSiblings = true,
): DocumentDetail {
  const rows = conn
    .prepare(`SELECT id FROM documents WHERE id LIKE ? ESCAPE '${LIKE_ESCAPE}'`)
    .all(escapeLikePrefix(idOrPrefix)) as { id: string }[];
  if (rows.length === 0) {
    throw new DocumentNotFoundError(`no document with id (prefix) ${JSON.stringify(idOrPrefix)}`);
  }
  if (rows.length > 1) {
    throw new AmbiguousDocumentPrefixError(
      `ambiguous prefix ${JSON.stringify(idOrPrefix)} matches ${rows.length} documents`,
    );
  }
  const matchedId = (rows[0] as { id: string }).id;
  const hydrated = hydrate(conn, [[matchedId, 0.0, null]])[0];
  if (hydrated === undefined) {
    // Unreachable: `matchedId` was just read from `documents`, so `hydrate`
    // (which batches its own SELECT against the same table) cannot miss it.
    throw new DocumentNotFoundError(`no document with id (prefix) ${JSON.stringify(idOrPrefix)}`);
  }
  // Drop score/matched_text (search-only fields) from the get-by-id shape,
  // mirroring Python's `del doc["score"], doc["matched_text"]`.
  const { score: _score, matched_text: _matchedText, ...rest } = hydrated;
  const doc: DocumentDetail = { ...rest };
  if (!includeSiblings) {
    doc.siblings = [];
  }
  if (includeRaw) {
    if (vaultPath === undefined) {
      throw new Error("include_raw requires vault_path");
    }
    const filePath = join(vaultPath, doc.file_path);
    let text: string;
    try {
      text = readFileSync(filePath, "utf-8");
    } catch (e) {
      const err = e as NodeJS.ErrnoException;
      if (err.code === "ENOENT") {
        throw new DocumentFileMissing(
          `file moved or deleted since last ingest: ${JSON.stringify(doc.file_path)} ` +
            `(document ${JSON.stringify(doc.document_id)}) — re-run \`qkb ingest\``,
        );
      }
      // Below-the-cut: ENOENT is the common case (moved/deleted since last
      // ingest) and gets the friendly message above. Anything else
      // (EISDIR, EACCES, ...) still means "can't read this document's
      // file" and should surface as the same typed, catchable error rather
      // than a raw exception.
      throw new DocumentFileMissing(
        `cannot read file ${JSON.stringify(doc.file_path)} (document ${JSON.stringify(doc.document_id)}): ${err.message}`,
      );
    }
    const desc = contextDescription(conn, doc.context);
    doc.raw_text = desc ? `<!-- Context: ${desc} -->\n\n${text}` : text;
  }
  return doc;
}
